### ISSUE TITLE
Should we clarify the intent of the duplicate SAM flag 0x400 ?

### ISSUE THREAD

**Issue by yfarjoun** [samtools/hts-specs#445]
The spec is very silent about the meaning of the duplicate SAM flag. In particular, is it a RECORD flag or a TEMPLATE flag. 

While I don't think that we should codify how duplicates should or should not be marked (meaning by what algorithm), I think that we should recommend (or require) that all records from the same template (i.e. with the same queryname) have the same value for bit 0x400. 

This is currently implemented in Picard MarkDuplicates but _if_ the input is query-sorted. 

Other people's thoughts on the matter?

**Comment by lbergelson** [comment]
Are there any technologies that can produce duplicates of a single read rather than of the template?  I think a recommendation at least for Illumina data would make a lot of sense.  I don't think it should be illegal to have templates unevenly marked since I think reasonable people might decide that the trade off of marking individual reads vs pairs might be worthwhile in their use case.

**Comment by tfenne** [comment]
It probably would be good to clarify the flag in general.  The current description is pretty short!  E.g. it doesn't even state what the standard practice is if you have two reads that are duplicates of one another (mark one, mark both?)!

I don't think there should be any prescription that the flag should have the same value for all reads for a template though.  I'm hard pressed to come up with a good example off the cuff of when it would be preferred to mark one end and not the other, but a lot of implementations will for expediency's sake not mark all records for a template.  It then also opens up questions like: what if the reads have multiple mappings?  Should all be marked as duplicates or only those that exist at a location with duplicates?  Etc.

**Comment by d-cameron** [comment]
>It then also opens up questions like: what if the reads have multiple mappings? Should all be marked as duplicates or only those that exist at a location with duplicates? 

Depends on what sort of multiple mapping. Split read alignments? Yes. Alternate alignments? Probably - there's a stronger case for not doing it in this scenario. That said, we haven't even clarified whether the set of alignments for a template is a set of distinct (possibly compound) alignments, or a graph of partial template alignments so it's still not entirely clear what multiple mappings actually are (i.e.: if we have A-B and A-C chimeric alignments, do we write two A records?) 

I've had issues with inconsistent duplication marking causing downstream issues. In particular, the standard duplicate marking programs, by default, don't mark supplementary alignments. This means downstream filtering of duplicate records results in orphaned supplementary alignments.

This seems like something that would be a good fit for section 2 of the SAM specs, or the SAM strict sub-specifications #283 that I'd like to resurrect (and would/could replace section 2 of the specs).

**Comment by jkbonfield** [comment]
I'm hesitant to codify program behaviour here.  We could something to the recommended section maybe that is guidance but not normative specification, but we paint ourselves into corners if we write the spec with current paired Illumina libraries in mind.

As for whether it is a record or a template flag, surely it's a record one as every record has its own flag.  Now we can argue about what constitutes consistency between records, but applies for other flags too.   For example stating the next segment is unmapped and having the next segment not unmapped is something that "samtools fixmates" will fix, but it's not something the spec explains.  Perhaps that also needs clarifying, but generally I've viewed that sort of thing is best left implicit and validated by implementations.

**Comment by d-cameron** [comment]
Anything implicit might as well be undefined there is no guarantee that
other tools agree with your interpretation. I would argue this should be a
template level attribute just like the read paired flag. That is a record
level attribute since we have no way to define a template attribute other
than on records.

**Comment by jkbonfield** [comment]
I don't understand.  FLAGS *are* record level.  That's the spec; each and every record has its own FLAG value.

Now whether or not they are required to be consistent between all records on a template is a different question, although still important.  That's the bit I say is implicit right now.

It's basically the difference between syntax and semantics.  The SAM syntax is that flags are per record.  The semantics will be how we're interpreting them.  Traditionally we've focused more on the former than the latter.

**Comment by yfarjoun** [comment]
@jkbonfield Let's take a step back. everyone here is aware that technically the bits are record level, and that technically one **could** have different values on different records of the same template...that NOT the question.

I think that you will agree that some of the flags are actually template-level bits in that it makes no sense to have them disagree.
while it isn't written anywhere I would assert that 0x1,0x2 are of that kind:
   > template having multiple segments in sequencing

is a *template* level bit, which should the same value for all records stemming from the same template.
   > each segment properly aligned according to the aligner

should also be thought of as a template level bit, since otherwise there's no way to interpret the `each segment` part of the sentence. (especially in contrast to 0x4 which is simply `segment unmapped`.

So, perhaps this issue has grown to be an issue that discusses whether the spec should be explicit about the template-level agreement of record-level flags for various bits....

I think that we should be explicit about 0x1, 0x2, and 0x200. and I think that we should say something like "0x400 bit **must** have the same value for all aligned, primary records (FLAG & 0x904==0) from the same template, and recommended to be same for all records from the same template".

**Comment by jkbonfield** [comment]
I agree we can move beyond a format and into data semantics, but the wording is vital.  You have to be **very** careful.  For example "aligned, primary records (FLAG & 0x904==0)".  The spec defines primary as non-secondary, not as non-secondary and non-supplementary.  Thus a chimeric alignment can be primary along its entire alignment length.  The above statement introduces an ambiguity.  Sadly however I see we already have a statement on only one read having FLAG & 0x900 == 0 so an ambiguity is already there. :-(

I'd argue that it's better to keep format and meaning separated a little bit, maybe with some term like "well formed" or similar so we can claim this data is valid SAM format but internally not well-formed.  We have data sets where flags don't match up properly as historically there have been situations where aligners produced such data - maybe still do.  I'm not trying to defend them (it's a disgrace!) but we need to still be able to categorise their output in some manner; they're still SAM, even if it's SAM with inconsistencies.

As for what we should be stating when it comes to a well formed, consistent, SAM - agreed some things really are obvious and I take the point about some flags being constant along the template and others being constant along a specific alignment of that template.  1 and 2: agreed with you.  4/8 should be consistent between one alignment and the NEXT, as should 0x10/0x20.  0x40/0x80 - I guess we can ensure that only one read on a primary alignment has these set, although it could be both on the same read.  0x200 - disagree: there's no reason why vendor specific QC should be all or nothing; we may want to fail read 1 and pass read 2.    0x400 - mostly agree given it states "PCR or optical", both of which are template level things, but I wouldn't be surprised if someone had some other (wrong?) duplicate definition out there.

**Comment by jkbonfield** [comment]
> I think that we should be explicit about 0x1, 0x2, and 0x200. and I think that we should say something like "0x400 bit **must** have the same value for all aligned, primary records (FLAG & 0x904==0) from the same template, and recommended to be same for all records from the same template".

Also pointed out in the office - why "aligned"?  Is it not correct to say an unaligned record may also be a PCR duplicate?

The difficulty in getting this right is why I feel it's dangerous to cast these in stone as "the specification" and not a recommendation or some extra level of interpretation on top of the specification to produce something like "well-formed" or "validated" SAM.

**Comment by yfarjoun** [comment]
The only reason was laziness (meaning, accounting for the laziness of the implementation), it's easier to "find" the records when they are aligned, and so it is easier to mark them....I agree that in a perfect scenario, all the reads from the same template would have the same duplicate marking, which is what my next sentence tried to imply....

**Comment by jkbonfield** [comment]
> The only reason was laziness (meaning, accounting for the laziness of the implementation), it's easier to "find" the records when they are aligned, and so it is easier to mark them....I agree that in a perfect scenario, all the reads from the same template would have the same duplicate marking, which is what my next sentence tried to imply....

I don't like this.  It's misleading as the implicit implication is that the spec recommends duplicate detection only considers aligned data.  Someone may have another method which we don't want to reject.  Eg READ1 is mapped in two templates to identical location and READ2 in both templates are unmapped with identical sequence.  It's thus likely a duplicate template.

We can't get into such details in the spec however.  Hence we also shouldn't be getting into specific bit patterns that make sense only for certain implementations.  If we want it described, it should be agnostic: e.g. "All records forming an alignment of a template must share the same bit value of the 0x400 (DUP) FLAG".   It's tricky to define this well.  What if we have two possible alignments for a template, a primary and a secondary, with one exactly aligning against another template (PCR DUP) and the other not (not PCR-DUP?).

**Comment by yfarjoun** [comment]
I agree with your dislike. I'm trying to thread this needle....the tension is between "correctness" (a template is the thing that is duplicated, it is quite non-sensical if one record is a duplicate and another from the same template is not) and the implementations that exist in the wild (which, for the most part ignore unmapped, secondary and supplementary reads when marking dups). I think that the result will be some compromise...

In the case you describe, I think that the duplicate-marking implementation has to make a call: is it or is it not a duplicate? I don't want to tell implementations how to make the call, only how to record the call they make.

How about specifying how inconsistent marking should be **interpreted**? For example: "In the case that one or more records from a template is marked as duplicate, the entire template should be considered a duplicate" (to enable lazy/legacy implementations and tools to "fix" inconsistent markings)

**Comment by jkbonfield** [comment]
Agreed it's hard to thread the needle and I don't think the general premise of tying down the spec is flawed.  I'm just arguing for clearer distinction between syntax and meaning.  However it does appear that ship has already sailed so I'm starting to relent on this one. :-)

Marking how things are interpreted is an interesting difference, but open to the same flaws of bad wording.  Eg the above rules out specifying a primary template alignment (whole alignment; ie all primary records) and a secondary template alignment from having different DUP status.  I can see why some tools may be written to distinguish those cases.   (E.g. if we have two different places we could feasibly locate the read-pair, one of which dups and one of which doesn't.)

Rephrasing the above to "one of more records from a template **alignment**" would fix it.  If so we should probably add "template alignment" to the glossary at the start, defined to be a set of read alignments (either linear or chimeric) covering the template, with explicit permission to have multiple template alignments as that's important for considering how PNEXT/RNEXT work in secondary alignments.  (That's a whole different can of worms which I've given up trying to resolve sadly.)

**Comment by yfarjoun** [comment]
ugh....that makes it really complex...imagine the following scenario:
(all in chr1, ignoring strand, and assuming that duplicate marking is based on POS for simplicity)
read 1/1 POS=100, read 1/2 POS=300, read 1/1 _secondary_ POS=200
read2/1 POS=200, read 2/2 POS=300

if we look only at the primaries, read1/1+read1/2 are differently aligned than read2/1 and read2/2 so none of these reads should be marked as duplicates. But, if we look at the _secondary_ alignment of read1/1, it would seem that together with read1/2 they are a duplicate alignment to read2/1 and read2/2, so in that case perhaps read1/1 secondary and read1/2 should be marked as duplicates. 

In this case, how should the only alignment of read1/2 be marked? 

The way to resolve this is to have a secondary alignment of read1/2 which is identical to the primary alignment of read1/2...but I don't think that we have required this, nor put in place mechanism for explaining whats going on...

I think that just like in variant calling, the duplicate-marking algorithms need to make judgement calls, and how to mark templates in ambiguous cases is one of them. Kicking the can down the road doesn't actually make the problem go away....

**Comment by jkbonfield** [comment]
I'd say read 1/1 2nd and read 1/2 2nd as a pair are duplicate in that scenario.

I'm not saying this is easy and I agree algorithms need to make judgement calls.  However we need to block non-sensical things rather than trying to state how the current methods work with wording that assumes data is aligned or that all alignments for the template must be treated in unison.
